metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,234 @@
+namespace graph {
+DBGSSHash::~DBGSSHash() {}
+
+DBGSSHash::DBGSSHash(size_t k):k_(k) {
+    dict_ = std::make_unique<sshash::dictionary>();
+}
+
+DBGSSHash::DBGSSHash(std::string const& input_filename, size_t k):k_(k){
+    sshash::build_configuration build_config;
+    build_config.k = k;//
+    // quick fix for value of m... k/2 but odd
+    build_config.m = (k_+1)/2;
+    if(build_config.m % 2 == 0) build_config.m++;
+    dict_ = std::make_unique<sshash::dictionary>();
+    dict_->build(input_filename, build_config);
+}
+    std::string DBGSSHash::file_extension() const  { return kExtension; }
+    size_t DBGSSHash::get_k() const  { return k_; }
+    DeBruijnGraph::Mode DBGSSHash::get_mode() const  { return BASIC; }
+
+void DBGSSHash::add_sequence(std::string_view sequence,
+                             const std::function<void(node_index)> &on_insertion) {
+    // TODO: throw exception? :)
+        throw std::runtime_error("adding sequences not implemented");
+
+}
+
+void DBGSSHash::map_to_nodes(std::string_view sequence,
+                             const std::function<void(node_index)> &callback,
+                             const std::function<bool()> &terminate) const {
+    map_to_nodes_sequentially(sequence, callback, terminate);
+}
+
+void DBGSSHash ::map_to_nodes_sequentially(std::string_view sequence,
+                                           const std::function<void(node_index)> &callback,
+                                           const std::function<bool()> &terminate) const {
+    for (size_t i = 0; i + k_ <= sequence.size() && !terminate(); ++i) {
+        callback(kmer_to_node(sequence.substr(i, k_)));
+    }
+}
+
+DBGSSHash::node_index DBGSSHash::traverse(node_index node, char next_char) const {
+    std::string kmer = DBGSSHash::get_node_sequence(node); 
+    sshash::neighbourhood nb = dict_->kmer_forward_neighbours(&kmer[0]);
+    uint64_t ssh_idx = -1; 
+    switch (next_char) {
+    case 'A':
+        ssh_idx = nb.forward_A.kmer_id;
+        break;
+    case 'C':
+        ssh_idx = nb.forward_C.kmer_id;
+        break;
+    case 'G':
+        ssh_idx = nb.forward_G.kmer_id;
+        break;
+    case 'T':
+        ssh_idx = nb.forward_T.kmer_id;
+        break;
+    default:
+        break;
+    }
+    return ssh_idx + 1;
+}
+
+DBGSSHash::node_index DBGSSHash::traverse_back(node_index node, char prev_char) const {
+    std::string kmer = DBGSSHash::get_node_sequence(node);
+    sshash::neighbourhood nb = dict_->kmer_backward_neighbours(&kmer[0]);
+    uint64_t ssh_idx = -1; 
+    switch (prev_char) {
+    case 'A':
+        ssh_idx = nb.backward_A.kmer_id;
+        break;
+    case 'C':
+        ssh_idx = nb.backward_C.kmer_id;
+        break;
+    case 'G':
+        ssh_idx = nb.backward_G.kmer_id;
+        break;
+    case 'T':
+        ssh_idx = nb.backward_T.kmer_id;
+        break;
+    default:
+        break;
+    }
+    return ssh_idx + 1;
+}
+
+void DBGSSHash ::adjacent_outgoing_nodes(node_index node,
```

Comment:
The reason why we kept this implementation out of the parent class is because we wanted to give a hint to any potential users that `adjacent_outgoing_nodes` may be implemented more efficiently than `call_outgoing_kmers`, so we force them to implement them separately.

Do you think this is too much hand-holding?

Reply:
Yes, I'd say so. I think if the users really need it to be more efficient, they'll figure out that it should be overriden, and this argument kind of applies to any other virtual function with a default implementation anyway?